system-addon/lib/ASRouterTargeting.jsm

```diff
@@ -0,0 +1,52 @@
+ChromeUtils.import("resource://gre/modules/components-utils/FilterExpressions.jsm");
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+ChromeUtils.defineModuleGetter(this, "ProfileAge",
+  "resource://gre/modules/ProfileAge.jsm");
+ChromeUtils.import("resource://gre/modules/Console.jsm");
+
+const FXA_USERNAME_PREF = "services.sync.username";
+
+/**
+ * removeRandomItemFromArray - Removes a random item from the array and returns it.
+ *
+ * @param {Array} arr An array of items
+ * @returns one of the items in the array
+ */
+function removeRandomItemFromArray(arr) {
+  return arr.splice(Math.floor(Math.random() * arr.length), 1)[0];
+}
+
+const TargetingGetters = {
+  get profileAgeCreated() {
+    return new ProfileAge(null, null).created;
+  },
+  get profileAgeReset() {
+    return new ProfileAge(null, null).reset;
+  },
+  get hasFxAccount() {
+    return Services.prefs.prefHasUserValue(FXA_USERNAME_PREF);
+  }
+};
+
+this.ASRouterTargeting = {
+  Environment: TargetingGetters,
+
+  isMatch(filterExpression, context = this.Environment) {
+    return FilterExpressions.eval(filterExpression, context);
+  },
+
+  async findMatchingMessage(originalArray, context) {
+    const arrayOfItems = [...originalArray];
+    let match;
+    let candidate;
+    while (!match && arrayOfItems.length) {
+      candidate = removeRandomItemFromArray(arrayOfItems);
+      if (candidate && (!candidate.targeting || await this.isMatch(candidate.targeting, context))) {
```

Comment:
Why do we want to match if the candidate has no targeting?

Reply:
If there's no targeting, presumably that means we could show it in all conditions (otherwise it would never be shown)

Reply:
Ah ok. So in the current state that we're in, it's ok that the onboarding tour still gets added to the rotation of messages to show, since it doesn't have any targeting yet?